In [7]:
from nltk import word_tokenize
from nltk.stem import PorterStemmer

sentence = "Don't be fooled by the dark sounding name, Mr. Jone's Orphanage is as cheery as cheery goes for a pastry shop. " 

In [8]:
words = word_tokenize(sentence)

In [10]:
print(len(words))
print(words)

25
['do', "n't", 'be', 'fool', 'by', 'the', 'dark', 'sound', 'name', ',', 'mr.', 'jone', "'s", 'orphanag', 'is', 'as', 'cheeri', 'as', 'cheeri', 'goe', 'for', 'a', 'pastri', 'shop', '.']


In [45]:
from scipy import sparse

data = ([5,2,3])
row = ([2,2,3])
col = ([3,4,2])

print(sparse.coo_matrix((data, (row, col)), (5, 6)).toarray() )

[[0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 5 2 0]
 [0 0 3 0 0 0]
 [0 0 0 0 0 0]]


In [44]:
data = [5, 2, 3]
row_index = [2, 2, 3]
col = [3, 4, 2]

print(sparse.csr_matrix((data, col, row_index), (5, 6)).toarray() )

[[0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 5 2 0]
 [0 0 3 0 0 0]
 [0 0 0 0 0 0]]


In [74]:
data = ([1, 2, 4, 5, 2, 3])
row = ([0, 0, 2, 2, 2, 4])
row_index = ([0, 2, 2, 5, 5, 6])
col = ([0, 1, 2, 3, 4, 2])

print(sparse.csr_matrix((data, col, row_index),).toarray())

[[1 2 0 0 0]
 [0 0 0 0 0]
 [0 0 4 5 2]
 [0 0 0 0 0]
 [0 0 3 0 0]]


In [138]:
import pandas as pd

movies = pd.read_csv('./tmdb_5000_movies.csv')
movies[:3]

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...   2015-10-26   880674609   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   
2    148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500  
2                                   Spectre           6.3        4466

In [139]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4803 entries, 0 to 4802
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4803 non-null   int64  
 1   genres                4803 non-null   object 
 2   homepage              1712 non-null   object 
 3   id                    4803 non-null   int64  
 4   keywords              4803 non-null   object 
 5   original_language     4803 non-null   object 
 6   original_title        4803 non-null   object 
 7   overview              4800 non-null   object 
 8   popularity            4803 non-null   float64
 9   production_companies  4803 non-null   object 
 10  production_countries  4803 non-null   object 
 11  release_date          4802 non-null   object 
 12  revenue               4803 non-null   int64  
 13  runtime               4801 non-null   float64
 14  spoken_languages      4803 non-null   object 
 15  status               

4803개의 레코드와 20개의 피처로 구성되어 있는 모습이다. 영화 제목, 개요, 인기도 등 영화에 대한 다양한 메타 정보를 가지고 있는데 이중 콘텐츠 필터링 추천 분석에 사용할 주요 칼럼만 추출해 새롭게 DataFrame으로 만들자

추출할 주요 할럼은 genres(id, title, 영화의 여러 속성), vote_average(평균 평점), vote_count(평점 투표수), popularity(영화의 인기), keywork(영화를 설명하는 주요 문구), overview(영화 개요 설명)

In [140]:
# movies_df = movies.iloc[:, [3, 17, 1, 18, 19, 8, 4, 7]]
# movies_df

movies_df = movies[['id', 'title', 'genres', 'vote_average', 'vote_count', 'popularity', 'keywords', 'overview']]
movies_df

id                                     title  \
0      19995                                    Avatar   
1        285  Pirates of the Caribbean: At World's End   
2     206647                                   Spectre   
3      49026                     The Dark Knight Rises   
4      49529                               John Carter   
...      ...                                       ...   
4798    9367                               El Mariachi   
4799   72766                                 Newlyweds   
4800  231617                 Signed, Sealed, Delivered   
4801  126186                          Shanghai Calling   
4802   25975                         My Date with Drew   

                                                 genres  vote_average  \
0     [{"id": 28, "name": "Action"}, {"id": 12, "nam...           7.2   
1     [{"id": 12, "name": "Adventure"}, {"id": 14, "...           6.9   
2     [{"id": 28, "name": "Action"}, {"id": 12, "nam...           6.3   
3     [{"id": 28, "name": "Action"}, {"id": 80, "nam...           7.6   
4     [{"id": 28, "name": "Action"}, {"id": 12, "nam...           6.1   
...                                                 ...           ...   
4798  [{"id": 28, "name": "Action"}, {"id": 80, "nam...           6.6   
4799  [{"id": 35, "name": "Comedy"}, {"id": 10749, "...           5.9   
4800  [{"id": 35, "name": "Comedy"}, {"id": 18, "nam...           7.0   
4801                                                 []           5.7   
4802                [{"id": 99, "name": "Documentary"}]           6.3   

      vote_count  popularity  \
0          11800  150.437577   
1           4500  139.082615   
2           4466  107.376788   
3           9106  112.312950   
4           2124   43.926995   
...          ...         ...   
4798         238   14.269792   
4799           5    0.642552   
4800           6    1.444476   
4801           7    0.857008   
4802          16    1.929883   

                                               keywords  \
0     [{"id": 1463, "name": "culture clash"}, {"id":...   
1     [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2     [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3     [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4     [{"id": 818, "name": "based on novel"}, {"id":...   
...                                                 ...   
4798  [{"id": 5616, "name": "united states\u2013mexi...   
4799                                                 []   
4800  [{"id": 248, "name": "date"}, {"id": 699, "nam...   
4801                                                 []   
4802  [{"id": 1523, "name": "obsession"}, {"id": 224...   

                                               overview  
0     In the 22nd century, a paraplegic Marine is di...  
1     Captain Barbossa, long believed to be dead, ha...  
2     A cryptic message from Bond’s past sends him o...  
3     Following the death of District Attorney Harve...  
4     John Carter is a war-weary, former military ca...  
...                                                 ...  
4798  El Mariachi just wants to play his guitar and ...  
4799  A newlywed couple's honeymoon is upended by th...  
4800  "Signed, Sealed, Delivered" introduces a dedic...  
4801  When ambitious New York attorney Sam is sent t...  
4802  Ever since the second grade when he first saw ...  

[4803 rows x 8 columns]

In [141]:
movies_df['keywords'][0]

'[{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "space war"}, {"id": 3388, "name": "space colony"}, {"id": 3679, "name": "society"}, {"id": 3801, "name": "space travel"}, {"id": 9685, "name": "futuristic"}, {"id": 9840, "name": "romance"}, {"id": 9882, "name": "space"}, {"id": 9951, "name": "alien"}, {"id": 10148, "name": "tribe"}, {"id": 10158, "name": "alien planet"}, {"id": 10987, "name": "cgi"}, {"id": 11399, "name": "marine"}, {"id": 13065, "name": "soldier"}, {"id": 14643, "name": "battle"}, {"id": 14720, "name": "love affair"}, {"id": 165431, "name": "anti war"}, {"id": 193554, "name": "power relations"}, {"id": 206690, "name": "mind and soul"}, {"id": 209714, "name": "3d"}]'

`literal_eval`
문자열을 문자열이 의미하는 list [dict1, dict2] 객체로 만들수 있다. Series 객체의 apply()에 이 함수를 적용하면 문자열을 객체로 변환한다.
→ 딕셔너리 형태를 가진 '문자열'을 딕셔너리로 변환시켜준다

```
import ast

str_dict = "{'key_1': 'value_1', 'key_2': 'value_2'}"
# str_dict는 딕셔너리 형태를 가진 문자열임

real_dict = ast.literal_eval(str_dict)
# real_dict는 딕셔너리 형태로 변환되었기 때문에 key-value를 사용할 수 있음
```

In [142]:
from ast import literal_eval

# movies['genres']의 것이 apply안의 것에 영향을 받는다
movies_df['genres'] = movies_df['genres'].apply(literal_eval)
movies_df['keywords'] = movies_df['keywords'].apply(literal_eval)

C:\Users\admin\AppData\Local\Temp\ipykernel_7024\3550466579.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies_df['genres'] = movies_df['genres'].apply(literal_eval)
C:\Users\admin\AppData\Local\Temp\ipykernel_7024\3550466579.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies_df['keywords'] = movies_df['keywords'].apply(literal_eval)


In [143]:
movies_df['genres'][0]

[{'id': 28, 'name': 'Action'},
 {'id': 12, 'name': 'Adventure'},
 {'id': 14, 'name': 'Fantasy'},
 {'id': 878, 'name': 'Science Fiction'}]

In [144]:
# for i in movies['genres'][0] :
#     print(i['name'])
[i['name'] for i in movies_df['genres'][0]]

['Action', 'Adventure', 'Fantasy', 'Science Fiction']

In [145]:
movies_df['genres'] = movies_df['genres'].apply(lambda x : [i['name'] for i in x])
movies_df['keywords'] = movies_df['keywords'].apply(lambda x : [i['name'] for i in x])

C:\Users\admin\AppData\Local\Temp\ipykernel_7024\2899412189.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies_df['genres'] = movies_df['genres'].apply(lambda x : [i['name'] for i in x])
C:\Users\admin\AppData\Local\Temp\ipykernel_7024\2899412189.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies_df['keywords'] = movies_df['keywords'].apply(lambda x : [i['name'] for i in x])


In [146]:
movies_df['genres']

0       [Action, Adventure, Fantasy, Science Fiction]
1                        [Adventure, Fantasy, Action]
2                          [Action, Adventure, Crime]
3                    [Action, Crime, Drama, Thriller]
4                [Action, Adventure, Science Fiction]
                            ...                      
4798                        [Action, Crime, Thriller]
4799                                [Comedy, Romance]
4800               [Comedy, Drama, Romance, TV Movie]
4801                                               []
4802                                    [Documentary]
Name: genres, Length: 4803, dtype: object

In [147]:
movies_df['keywords']

0       [culture clash, future, space war, space colon...
1       [ocean, drug abuse, exotic island, east india ...
2       [spy, based on novel, secret agent, sequel, mi...
3       [dc comics, crime fighter, terrorist, secret i...
4       [based on novel, mars, medallion, space travel...
                              ...                        
4798    [united states–mexico barrier, legs, arms, pap...
4799                                                   []
4800    [date, love at first sight, narration, investi...
4801                                                   []
4802            [obsession, camcorder, crush, dream girl]
Name: keywords, Length: 4803, dtype: object

### 장르 콘텐츠 유사도 측정
현재 여러개의 개별장르가 리스트로 구성되어 있는 genres 칼럼으로 장르별 유사도를 측정하는 가장 간단한 방법은 무엇일까? 아마 문자열로 변경한 뒤 이를 CountVectorizer로 피처 벡터화 한 후, 행렬 데이터 값을 코사인 유사도록 비교하는 것일 것이다.

리스트 객체 내의 개별 값을 연속된 문자열로 변환하려면 일반적으로 `'구분문자'.join(리스트 객체)`를 사용하면 된다.

In [148]:
# CountVectorizer를 적용하기 위해 공백 문자로 word 단위가 구분되는 문자열로 반환
movies_df['genres_literal'] = movies_df['genres'].apply(lambda x: ' '.join(x))
movies_df['genres_literal']

C:\Users\admin\AppData\Local\Temp\ipykernel_7024\1382422869.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies_df['genres_literal'] = movies_df['genres'].apply(lambda x: ' '.join(x))


0       Action Adventure Fantasy Science Fiction
1                       Adventure Fantasy Action
2                         Action Adventure Crime
3                    Action Crime Drama Thriller
4               Action Adventure Science Fiction
                          ...                   
4798                       Action Crime Thriller
4799                              Comedy Romance
4800               Comedy Drama Romance TV Movie
4801                                            
4802                                 Documentary
Name: genres_literal, Length: 4803, dtype: object

CountVectorizer 로 변환

In [149]:
from sklearn.feature_extraction.text import CountVectorizer

count_vect = CountVectorizer(min_df=0, ngram_range=(1, 2))
genre_mat = count_vect.fit_transform(movies_df['genres_literal'])

print(genre_mat.shape)

(4803, 276)


사이킷런의 `cosine_similarity()` 함수는 기준행과 비교행의 코사인 유사도를 행렬 형태로 반환하는 함수이다. 피처 벡터화된 행렬에 아래와 같이 적용해 볼 수 있다.

In [150]:
from sklearn.metrics.pairwise import cosine_similarity

genre_sim = cosine_similarity(genre_mat, genre_mat)
print(genre_sim.shape)
print(genre_sim[:2])

(4803, 4803)
[[1.         0.59628479 0.4472136  ... 0.         0.         0.        ]
 [0.59628479 1.         0.4        ... 0.         0.         0.        ]]


genre_mat을 기준으로 genre_mat을 비교했기 때문에 하나하나 전체를 비교해준다.

cosine_similarity 호출로 생성된 genre_sim 객체는 movies_df이 genre_literal 칼럼을 피처 벡터화할 행렬genre_mat) 데이터의 행(레코드)별 유사도를 가지고 있으며, 이는 결국 movies_df DataFrame의 행별 장르 유사도를 가지고 있는 것이 된다.

이제 기준 행별로 비교 대상이 되는 행의 유사도 값이 높은 순으로 정렬된 행렬의 위치 인덱스 값을 추출해보자. 값이 높은 순으로 정렬된 비교 대상 행의 유사도 '값'이 아니라 비교대상 행의 '위치 인덱스'임을 또 한번 주의하자. 넘파이의 `argsort()[:, ::-1]`함수를 이용하면 유사도가 높은 순으로 정리된 `genre_sim` 객체의 비교 행 위치 인덱스 값을 가져온다
```python
a = np.array([11, 10, 13, 25])
np.argsort(a)
```
```
array([1, 0, 2, 3], dtype=int64)
```

In [151]:
import numpy as np

np.argsort(genre_sim)[:,::-1]

array([[   0, 3494,  813, ..., 3038, 3037, 2401],
       [ 262,    1,  129, ..., 3069, 3067, 2401],
       [   2, 1740, 1542, ..., 3000, 2999, 2401],
       ...,
       [4800, 3809, 1895, ..., 2229, 2230,    0],
       [4802, 1594, 1596, ..., 3204, 3205,    0],
       [4802, 4710, 4521, ..., 3140, 3141,    0]], dtype=int64)

In [152]:
genre_sim_sorted = genre_sim.argsort()[:, ::-1][:1]
genre_sim_sorted

array([[   0, 3494,  813, ..., 3038, 3037, 2401]], dtype=int64)

In [153]:
def find_sim_movie(df, sorted_ind, title_name, top_n=10):
    
    # 인자로 입력된 movies_df DataFrame에서 'title' 컬럼이 입력된 title_name 값인 DataFrame추출
    title_movie = df[df['title'] == title_name]
    
    # title_named을 가진 DataFrame의 index 객체를 ndarray로 반환하고 
    # sorted_ind 인자로 입력된 genre_sim_sorted_ind 객체에서 유사도 순으로 top_n 개의 index 추출
    title_index = title_movie.index.values
    similar_indexes = sorted_ind[title_index, :(top_n)]
    
    # 추출된 top_n index들 출력. top_n index는 2차원 데이터 임. 
    #dataframe에서 index로 사용하기 위해서 1차원 array로 변경
    print(similar_indexes)
    similar_indexes = similar_indexes.reshape(-1)
    
    return df.iloc[similar_indexes] 

In [154]:
similar_movies = find_sim_movie(movies_df, genre_sim_sorted_ind, 'The Godfater', 10)
similar_movies[['title', 'vote_average']]

[]


Empty DataFrame
Columns: [title, vote_average]
Index: []

### 장르 콘텐츠 필터링을 이용한 영화 추천


In [130]:
C = movies_df['vote_average'].mean()
m = movies_df['vote_count'].quantile(0.6)
print(f'C : {round(C, 3)}, m : {round(m, 3)}')

C : 6.092, m : 370.2


In [131]:
percentile = 0.6
def weighted_vote_average(record) :
    v = record['vote_count']
    R = record['vote_average']
    
    return ((v/(v+m))*R) + ((m/(m+v))*C)

movies_df['weighted_vote'] = movies_df.apply(weighted_vote_average, axis=1)

C:\Users\admin\AppData\Local\Temp\ipykernel_7024\395683170.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies_df['weighted_vote'] = movies_df.apply(weighted_vote_average, axis=1)


In [132]:
movies_df['weighted_vote']

0       7.166301
1       6.838594
2       6.284091
3       7.541095
4       6.098838
          ...   
4798    6.290894
4799    6.089611
4800    6.106650
4801    6.084894
4802    6.100782
Name: weighted_vote, Length: 4803, dtype: float64

In [134]:
movies_df[['title', 'vote_average', 'weighted_vote', 'vote_count']]\
            .sort_values('weighted_vote', ascending=False)[:10]

title  vote_average  weighted_vote  vote_count
1881  The Shawshank Redemption           8.5       8.396052        8205
3337             The Godfather           8.4       8.263591        5893
662                 Fight Club           8.3       8.216455        9413
3232              Pulp Fiction           8.3       8.207102        8428
65             The Dark Knight           8.2       8.136930       12002
1818          Schindler's List           8.3       8.126069        4329
3865                  Whiplash           8.3       8.123248        4254
809               Forrest Gump           8.2       8.105954        7927
2294             Spirited Away           8.3       8.105867        3840
2731    The Godfather: Part II           8.3       8.079586        3338

In [136]:
def find_sim_movie(df, sorted_ind, title_name, top_n=10):
    title_movie = df[df['title'] == title_name]
    title_index = title_movie.index.values
    
    # top_n의 2배에 해당하는 쟝르 유사성이 높은 index 추출 
    similar_indexes = sorted_ind[title_index, :(top_n*2)]
    similar_indexes = similar_indexes.reshape(-1)
    
    # 기준 영화 index는 제외
    similar_indexes = similar_indexes[similar_indexes != title_index]
    
    # top_n의 2배에 해당하는 후보군에서 weighted_vote 높은 순으로 top_n 만큼 추출 
    return df.iloc[similar_indexes].sort_values('weighted_vote', ascending=False)[:top_n]

genre_sim_sorted_ind = genre_sim.argsort()[:, ::-1]
similar_movies = find_sim_movie(movies_df, genre_sim_sorted_ind, 'The Godfather',10)
similar_movies[['title', 'vote_average', 'weighted_vote']] 

title  vote_average  weighted_vote
2731       The Godfather: Part II           8.3       8.079586
1847                   GoodFellas           8.2       7.976937
3866                  City of God           8.1       7.759693
1663  Once Upon a Time in America           8.2       7.657811
883           Catch Me If You Can           7.7       7.557097
281             American Gangster           7.4       7.141396
4041              This Is England           7.4       6.739664
1149              American Hustle           6.8       6.717525
1243                 Mean Streets           7.2       6.626569
2839                     Rounders           6.9       6.530427

In [20]:
import pandas as pd
import numpy as np

movies = pd.read_csv('./ml-latest-small/ml-latest-small/movies.csv')
ratings = pd.read_csv('./ml-latest-small/ml-latest-small/ratings.csv')

In [21]:
movies[:3]

movieId                    title  \
0        1         Toy Story (1995)   
1        2           Jumanji (1995)   
2        3  Grumpier Old Men (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance

In [22]:
ratings[:3]

userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224

In [23]:
ratings = ratings[['userId', 'movieId', 'rating']]
ratings_matrix = ratings.pivot_table('rating', index='userId', columns='movieId')
ratings_matrix[:3]

movieId  1       2       3       4       5       6       7       8       \
userId                                                                    
1           4.0     NaN     4.0     NaN     NaN     4.0     NaN     NaN   
2           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  9       10      ...  193565  193567  193571  193573  193579  193581  \
userId                   ...                                                   
1           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  193583  193585  193587  193609  
userId                                   
1           NaN     NaN     NaN     NaN  
2           NaN     NaN     NaN     NaN  
3           NaN     NaN     NaN     NaN  

[3 rows x 9724 columns]

In [25]:
# title 칼럼을 얻기 위해 movies와 조인
rating_movies = pd.merge(ratings, movies, on='movieId')

# columns='title'로 피벗 수행
ratings_matrix = rating_movies.pivot_table('rating', index='userId', columns='title')

# NaN 값을 모두 0으로 변환
ratings_matrix = ratings_matrix.fillna(0)
ratings_matrix[:3]

title   '71 (2014)  'Hellboy': The Seeds of Creation (2004)  \
userId                                                        
1              0.0                                      0.0   
2              0.0                                      0.0   
3              0.0                                      0.0   

title   'Round Midnight (1986)  'Salem's Lot (2004)  \
userId                                                
1                          0.0                  0.0   
2                          0.0                  0.0   
3                          0.0                  0.0   

title   'Til There Was You (1997)  'Tis the Season for Love (2015)  \
userId                                                               
1                             0.0                              0.0   
2                             0.0                              0.0   
3                             0.0                              0.0   

title   'burbs, The (1989)  'night Mother (1986)  (500) Days of Summer (2009)  \
userId                                                                          
1                      0.0                   0.0                          0.0   
2                      0.0                   0.0                          0.0   
3                      0.0                   0.0                          0.0   

title   *batteries not included (1987)  ...  Zulu (2013)  [REC] (2007)  \
userId                                  ...                              
1                                  0.0  ...          0.0           0.0   
2                                  0.0  ...          0.0           0.0   
3                                  0.0  ...          0.0           0.0   

title   [REC]² (2009)  [REC]³ 3 Génesis (2012)  \
userId                                           
1                 0.0                      0.0   
2                 0.0                      0.0   
3                 0.0                      0.0   

title   anohana: The Flower We Saw That Day - The Movie (2013)  \
userId                                                           
1                                                     0.0        
2                                                     0.0        
3                                                     0.0        

title   eXistenZ (1999)  xXx (2002)  xXx: State of the Union (2005)  \
userId                                                                
1                   0.0         0.0                             0.0   
2                   0.0         0.0                             0.0   
3                   0.0         0.0                             0.0   

title   ¡Three Amigos! (1986)  À nous la liberté (Freedom for Us) (1931)  
userId                                                                    
1                         4.0                                        0.0  
2                         0.0                                        0.0  
3                         0.0                                        0.0  

[3 rows x 9719 columns]

In [27]:
ratings_matrix = ratings_matrix.T

In [29]:
ratings_matrix

userId                                     1    2    3    4    5    6    7    \
title                                                                          
'71 (2014)                                 0.0  0.0  0.0  0.0  0.0  0.0  0.0   
'Hellboy': The Seeds of Creation (2004)    0.0  0.0  0.0  0.0  0.0  0.0  0.0   
'Round Midnight (1986)                     0.0  0.0  0.0  0.0  0.0  0.0  0.0   
'Salem's Lot (2004)                        0.0  0.0  0.0  0.0  0.0  0.0  0.0   
'Til There Was You (1997)                  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
...                                        ...  ...  ...  ...  ...  ...  ...   
eXistenZ (1999)                            0.0  0.0  0.0  0.0  0.0  0.0  0.0   
xXx (2002)                                 0.0  0.0  0.0  0.0  0.0  0.0  0.0   
xXx: State of the Union (2005)             0.0  0.0  0.0  0.0  0.0  0.0  0.0   
¡Three Amigos! (1986)                      4.0  0.0  0.0  0.0  0.0  0.0  0.0   
À nous la liberté (Freedom for Us) (1931)  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

userId                                     8    9    10   ...  601  602  603  \
title                                                     ...                  
'71 (2014)                                 0.0  0.0  0.0  ...  0.0  0.0  0.0   
'Hellboy': The Seeds of Creation (2004)    0.0  0.0  0.0  ...  0.0  0.0  0.0   
'Round Midnight (1986)                     0.0  0.0  0.0  ...  0.0  0.0  0.0   
'Salem's Lot (2004)                        0.0  0.0  0.0  ...  0.0  0.0  0.0   
'Til There Was You (1997)                  0.0  0.0  0.0  ...  0.0  0.0  0.0   
...                                        ...  ...  ...  ...  ...  ...  ...   
eXistenZ (1999)                            0.0  0.0  0.0  ...  0.0  0.0  5.0   
xXx (2002)                                 0.0  1.0  0.0  ...  0.0  0.0  0.0   
xXx: State of the Union (2005)             0.0  0.0  0.0  ...  0.0  0.0  0.0   
¡Three Amigos! (1986)                      0.0  0.0  0.0  ...  0.0  0.0  0.0   
À nous la liberté (Freedom for Us) (1931)  0.0  0.0  0.0  ...  0.0  0.0  0.0   

userId                                     604  605  606  607  608  609  610  
title                                                                         
'71 (2014)                                 0.0  0.0  0.0  0.0  0.0  0.0  4.0  
'Hellboy': The Seeds of Creation (2004)    0.0  0.0  0.0  0.0  0.0  0.0  0.0  
'Round Midnight (1986)                     0.0  0.0  0.0  0.0  0.0  0.0  0.0  
'Salem's Lot (2004)                        0.0  0.0  0.0  0.0  0.0  0.0  0.0  
'Til There Was You (1997)                  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
...                                        ...  ...  ...  ...  ...  ...  ...  
eXistenZ (1999)                            0.0  0.0  0.0  0.0  4.5  0.0  0.0  
xXx (2002)                                 0.0  0.0  0.0  0.0  3.5  0.0  2.0  
xXx: State of the Union (2005)             0.0  0.0  0.0  0.0  0.0  0.0  1.5  
¡Three Amigos! (1986)                      0.0  0.0  0.0  0.0  0.0  0.0  0.0  
À nous la liberté (Freedom for Us) (1931)  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[9719 rows x 610 columns]

In [32]:
from sklearn.metrics.pairwise import cosine_similarity

item_sim = cosine_similarity(ratings_matrix, ratings_matrix)
item_sim_df = pd.DataFrame(data=item_sim, index=rating_matrix.columns, columns=rating_matrix.columns)

print(item_sim_df.shape)

(9719, 9719)


In [33]:
item_sim_df[:3]

title                                    '71 (2014)  \
title                                                 
'71 (2014)                                      1.0   
'Hellboy': The Seeds of Creation (2004)         0.0   
'Round Midnight (1986)                          0.0   

title                                    'Hellboy': The Seeds of Creation (2004)  \
title                                                                              
'71 (2014)                                                              0.000000   
'Hellboy': The Seeds of Creation (2004)                                 1.000000   
'Round Midnight (1986)                                                  0.707107   

title                                    'Round Midnight (1986)  \
title                                                             
'71 (2014)                                             0.000000   
'Hellboy': The Seeds of Creation (2004)                0.707107   
'Round Midnight (1986)                                 1.000000   

title                                    'Salem's Lot (2004)  \
title                                                          
'71 (2014)                                               0.0   
'Hellboy': The Seeds of Creation (2004)                  0.0   
'Round Midnight (1986)                                   0.0   

title                                    'Til There Was You (1997)  \
title                                                                
'71 (2014)                                                     0.0   
'Hellboy': The Seeds of Creation (2004)                        0.0   
'Round Midnight (1986)                                         0.0   

title                                    'Tis the Season for Love (2015)  \
title                                                                      
'71 (2014)                                                           0.0   
'Hellboy': The Seeds of Creation (2004)                              0.0   
'Round Midnight (1986)                                               0.0   

title                                    'burbs, The (1989)  \
title                                                         
'71 (2014)                                         0.000000   
'Hellboy': The Seeds of Creation (2004)            0.000000   
'Round Midnight (1986)                             0.176777   

title                                    'night Mother (1986)  \
title                                                           
'71 (2014)                                                0.0   
'Hellboy': The Seeds of Creation (2004)                   0.0   
'Round Midnight (1986)                                    0.0   

title                                    (500) Days of Summer (2009)  \
title                                                                  
'71 (2014)                                                  0.141653   
'Hellboy': The Seeds of Creation (2004)                     0.000000   
'Round Midnight (1986)                                      0.000000   

title                                    *batteries not included (1987)  ...  \
title                                                                    ...   
'71 (2014)                                                          0.0  ...   
'Hellboy': The Seeds of Creation (2004)                             0.0  ...   
'Round Midnight (1986)                                              0.0  ...   

title                                    Zulu (2013)  [REC] (2007)  \
title                                                                
'71 (2014)                                       0.0      0.342055   
'Hellboy': The Seeds of Creation (2004)          0.0      0.000000   
'Round Midnight (1986)                           0.0      0.000000   

title                                    [REC]² (2009)  \
title                                                    
'71 (2014)                                 

In [36]:
item_sim_df["Godfather, The (1972)"].sort_values(ascending=False)[1:7]

title
Godfather: Part II, The (1974)                           0.821773
Goodfellas (1990)                                        0.664841
One Flew Over the Cuckoo's Nest (1975)                   0.620536
Star Wars: Episode IV - A New Hope (1977)                0.595317
Fargo (1996)                                             0.588614
Star Wars: Episode V - The Empire Strikes Back (1980)    0.586030
Name: Godfather, The (1972), dtype: float64

In [48]:
def predict_rating(ratings_arr, item_sim_arr) : 
    ratings_pred = ratings_arr.dot(item_sim_arr)/np.array([np.abs(item_sim_arr).sum(axis=1)])
    return ratings_pred

In [49]:
ratings_matrix = ratings_matrix.T
ratigns_pred = predict_rating(ratings_matrix.values, item_sim_df.values)
ratings_pred_matrix = pd.DataFrame(data=ratigns_pred, index=rating_matrix.index, columns = rating_matrix.columns)
ratings_pred_matrix[:3]

title   '71 (2014)  'Hellboy': The Seeds of Creation (2004)  \
userId                                                        
1         0.070345                                 0.577855   
2         0.018260                                 0.042744   
3         0.011884                                 0.030279   

title   'Round Midnight (1986)  'Salem's Lot (2004)  \
userId                                                
1                     0.321696             0.227055   
2                     0.018861             0.000000   
3                     0.064437             0.003762   

title   'Til There Was You (1997)  'Tis the Season for Love (2015)  \
userId                                                               
1                        0.206958                         0.194615   
2                        0.000000                         0.035995   
3                        0.003749                         0.002722   

title   'burbs, The (1989)  'night Mother (1986)  (500) Days of Summer (2009)  \
userId                                                                          
1                 0.249883              0.102542                     0.157084   
2                 0.013413              0.002314                     0.032213   
3                 0.014625              0.002085                     0.005666   

title   *batteries not included (1987)  ...  Zulu (2013)  [REC] (2007)  \
userId                                  ...                              
1                             0.178197  ...     0.113608      0.181738   
2                             0.014863  ...     0.015640      0.020855   
3                             0.006272  ...     0.006923      0.011665   

title   [REC]² (2009)  [REC]³ 3 Génesis (2012)  \
userId                                           
1            0.133962                 0.128574   
2            0.020119                 0.015745   
3            0.011800                 0.012225   

title   anohana: The Flower We Saw That Day - The Movie (2013)  \
userId                                                           
1                                                0.006179        
2                                                0.049983        
3                                                0.000000        

title   eXistenZ (1999)  xXx (2002)  xXx: State of the Union (2005)  \
userId                                                                
1              0.212070    0.192921                        0.136024   
2              0.014876    0.021616                        0.024528   
3              0.008194    0.007017                        0.009229   

title   ¡Three Amigos! (1986)  À nous la liberté (Freedom for Us) (1931)  
userId                                                                    
1                    0.292955                                   0.720347  
2                    0.017563                                   0.000000  
3                    0.010420                                   0.084501  

[3 rows x 9719 columns]

In [51]:
from sklearn.metrics import mean_squared_error

# 사용자가 평점을 부여한 영화에 대해서만 예측 성능 평가 MSE를 구할 것이다
def get_mse(pred, actual) :
    # 평점이 있는 실제 영화만 추출
    pred = pred[actual.nonzero()].flatten()
    actual = actual[actual.nonzero()].flatten()
    
    return mean_squared_error(pred, actual)

MSE1 = get_mse(ratigns_pred, ratings_matrix.values)
print(f'아이템 기반 모든 인접 이웃 MSE: {MSE1:.4f}')

아이템 기반 모든 인접 이웃 MSE: 9.8954


In [52]:
def predict_rating_topsim(ratings_arr, item_sim_arr, N=20):
    # 사용자-아이템 평점 행렬 크기만큼 0으로 채운 예측 행렬 초기화
    pred = np.zeros(ratings_arr.shape)

    # 사용자-아이템 평점 행렬의 열 크기(아이템 수)만큼 반복 (row: 사용자, col: 아이템)
    for col in range(ratings_arr.shape[1]):
        
        # 특정 아이템의 유사도 행렬 오름차순 정렬시 index
        # 를 역순으로 나열시 상위 N개의 index = 특정 아이템의 유사도 상위 N개 아이템 인덱스 반환
        temp = np.argsort(item_sim_arr[:, col]) 
        top_n_items = [ temp[:-1-N:-1] ]
        
        # 개인화된 예측 평점을 계산: 반복당 특정 아이템의 예측 평점(사용자 전체)
        for row in range(ratings_arr.shape[0]):
            # 유사도 행렬
            item_sim_arr_topN = item_sim_arr[col, :][top_n_items].T # N x 1
            
            # 실제 평점 행렬
            ratings_arr_topN = ratings_arr[row, :][top_n_items]     # 1 x N
            
            # 예측 평점
            pred[row, col] = ratings_arr_topN @ item_sim_arr_topN
            pred[row, col] /= np.sum( np.abs(item_sim_arr_topN) )
            
    return pred

In [53]:
# 사용자별 예측 평점
ratings_pred = predict_rating_topsim(ratings_matrix.values , item_sim_df.values, N=20)

# 성능 평가
MSE2 = get_mse(ratings_pred, ratings_matrix.values )
print(f'아이템 기반 인접 TOP-20 이웃 MSE: {MSE2:.4f}')

# 예측 평점 데이터 프레임
ratings_pred_matrix = pd.DataFrame(data=ratigns_pred, index= ratings_matrix.index,
                                   columns = ratings_matrix.columns)

C:\Users\admin\AppData\Local\Temp\ipykernel_8424\1868812196.py:16: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  item_sim_arr_topN = item_sim_arr[col, :][top_n_items].T # N x 1
C:\Users\admin\AppData\Local\Temp\ipykernel_8424\1868812196.py:19: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  ratings_arr_topN = ratings_arr[row, :][top_n_items]     # 1 x N


아이템 기반 인접 TOP-20 이웃 MSE: 3.6950


In [54]:
# userId 9가 높은 평점을 준 영화 (실제 평점)
user_rating_id = ratings_matrix.loc[9, :]
user_rating_id[ user_rating_id > 0].sort_values(ascending=False)[:10]

title
Adaptation (2002)                                                                 5.0
Citizen Kane (1941)                                                               5.0
Raiders of the Lost Ark (Indiana Jones and the Raiders of the Lost Ark) (1981)    5.0
Producers, The (1968)                                                             5.0
Lord of the Rings: The Two Towers, The (2002)                                     5.0
Lord of the Rings: The Fellowship of the Ring, The (2001)                         5.0
Back to the Future (1985)                                                         5.0
Austin Powers in Goldmember (2002)                                                5.0
Minority Report (2002)                                                            4.0
Witness (1985)                                                                    4.0
Name: 9, dtype: float64

In [55]:
# 아직 보지 않은 영화 리스트 함수
def get_unseen_movies(ratings_matrix, userId):
    
    # user_rating: userId의 아이템 평점 정보 (시리즈 형태: title을 index로 가진다.)
    user_rating = ratings_matrix.loc[userId,:]
    
    # user_rating=0인 아직 안본 영화
    unseen_movie_list = user_rating[ user_rating == 0].index.tolist()
    
    # 모든 영화명을 list 객체로 만듬. 
    movies_list = ratings_matrix.columns.tolist()
    
    # 한줄 for + if문으로 안본 영화 리스트 생성
    unseen_list = [ movie for movie in movies_list if movie in unseen_movie_list]
    
    return unseen_list


# 보지 않은 영화 중 예측 높은 순서로 시리즈 반환
def recomm_movie_by_userid(pred_df, userId, unseen_list, top_n=10):    
    recomm_movies = pred_df.loc[userId, unseen_list].sort_values(ascending=False)[:top_n]
    
    return recomm_movies

In [56]:
# 아직 보지 않은 영화 리스트
unseen_list = get_unseen_movies(ratings_matrix, 9)

# 아이템 기반의 최근접 이웃 협업 필터링으로 영화 추천
recomm_movies = recomm_movie_by_userid(ratings_pred_matrix, 9, unseen_list, top_n=10)

# 데이터 프레임 생성
recomm_movies = pd.DataFrame(data=recomm_movies.values, index=recomm_movies.index, columns=['pred_score'])
recomm_movies

pred_score
title                                                         
Venom (1982)                                          0.303278
Dr. Goldfoot and the Bikini Machine (1965)            0.258705
Frankie and Johnny (1966)                             0.234754
English Vinglish (2012)                               0.214774
Harmonists, The (1997)                                0.169338
Passenger, The (Professione: reporter) (1975)         0.163884
Marriage of Maria Braun, The (Ehe der Maria Bra...    0.163884
Child, The (L'enfant) (2005)                          0.163884
3:10 to Yuma (1957)                                   0.163884
Story of Women (Affaire de femmes, Une) (1988)        0.163884

파이썬 기반의 추천 시스템 패키지인 Surprise은 다양한 추천 알고리즘을 쉽게 적용해 추천 시스템을 구축할 수 있다.

In [2]:
import surprise
print(surprise.__version__)

1.1.1


In [3]:
from surprise import Dataset

# ml-100k: 10만 개 평점 데이터
data = Dataset.load_builtin('ml-100k')

Dataset ml-100k could not be found. Do you want to download it? [Y/n] 

 y


Trying to download dataset from http://files.grouplens.org/datasets/movielens/ml-100k.zip...
Done! Dataset ml-100k has been saved to C:\Users\admin/.surprise_data/ml-100k


In [4]:
from surprise.model_selection import train_test_split

# surprise의 train_test_split() 사용
trainset, testset = train_test_split(data, test_size=0.25, random_state=0)

SVD로 잠재 요인 협업 필터링을 수행하겠다.

In [6]:
from surprise import SVD

# 알고리즘 객체 생성
algo = SVD(random_state=0)

# 이 객체에 fit을 수행해 학습 데이터 세트 기반으로 추천 알고리즘 학습
algo.fit(trainset)

학습된 추천 알고리즘을 기반으로 테스트 데이터 세트에 대해 추천을 수행. Surprise 에서 추천을 예측하는 메서드는 test(), predict() 두 개.

test() : 사용자-아이템 평점 데이터 세트 전체에 대해서 추천을 예측하는 메서드
         즉, 입력된 데이터 세트에 대해 추천 데이터 세트를 만들어줌
predict() : 개별 사용자와 영화에 대한 추천 평점을 반환

In [7]:
preds = algo.test(testset)

print(preds[0])
print(f'prediction type : {type(preds)}, size : {len(preds)}')

user: 120        item: 282        r_ui = 4.00   est = 3.51   {'was_impossible': False}
prediction type : <class 'list'>, size : 25000


SVD 알고리즘 객체의 test 메서드의 호출 결과는 파이썬 리스트이며 크기는 입력인자 데이터 세트의 크기와 같은 25000개인것을 확인. 리스트 내에 내포된 `Prediction` 객체의 uid, iid, r_ui, est 등의 속성에 접근하려면 객체명.uid와 같은 형식으로 가능하다. 추출해보자

In [9]:
[(pred.uid, pred.iid, pred.est) for pred in preds[:5]]

[('120', '282', 3.5114147666251547),
 ('882', '291', 3.573872419581491),
 ('535', '507', 4.033583485472447),
 ('697', '244', 3.8463639495936905),
 ('751', '385', 3.1807542478219157)]

이번에는 Surprise 의 다른 추천 예측 메서드인 predict(0를 이용해 추천 예측을 해보려 한다. predict()는 개별 사용자의 아이템에 대한 추천 평점을 예측해준다. 인자로 개별 사용자 아이디, 아이템 아이디를 입력하면 추천 예측 평점을 포함한 정보를 반환하는데, 이때 문자열로 들어가야한다.

In [11]:
uid = str(196)
iid = str(302)
pred = algo.predict(uid, iid)
print(pred)

user: 196        item: 302        r_ui = None   est = 4.49   {'was_impossible': False}


결과처럼 predict()는 개별 사용자와 아이템 정보를 받아 추천 예측 평점을 est로 반환하고 있다. 이제 테스트 데이터 셋을 이용해 추천 예측 평점과 실제 평점 사이의 차이를 평가해보려 한다.

In [12]:
from surprise import accuracy

accuracy.rmse(preds)

RMSE: 0.9467


0.9466860806937948

In [17]:
import pandas as pd

ratings = pd.read_csv('./ml-latest-small/ml-latest-small/ratings.csv')
# 위를 파일로 언로드시 인덱스와 헤더를 모두 제거한 새로운 파일 생성
ratings.to_csv('ratings_noheader', index=False, header=False)

In [19]:
from surprise import Reader

reader = Reader(line_format='user item rating timestamp', sep=', ', rating_scale=(0.5, 5))
data = Dataset.load_from_file('ratings_noheader.csv', reader=reader)

FileNotFoundError: [Errno 2] No such file or directory: 'ratings_noheader.csv'